cmd/zoekt-sourcegraph-indexserver/backoff_test.go

```diff
@@ -0,0 +1,396 @@
+	failedCount := 5
+	backoffDuration := 1 * time.Minute
+	maxBackoffDuration := time.Duration(failedCount) * backoffDuration
+	opts := IndexOptions{RepoID: 1, Name: "foo"}
+
+	backoff := backoff{
+		backoffDuration: backoffDuration,
+		maxBackoff:      maxBackoffDuration,
+	}
+
+	now := time.Now()
+	expectedFailuresCount := 0
+
+	for i := 0; i < failedCount; i++ {
+		backoff.Fail(now.Add(time.Duration(i)*backoffDuration), logtest.Scoped(t), opts)
+		expectedFailuresCount++
+		AssertFailuresCount(t, expectedFailuresCount, backoff)
+	}
+}
+
+func TestBackoff_MaximumConsecutiveFailures(t *testing.T) {
+	maximumCount := 3
+	failedCount := 2 * maximumCount
+	backoffDuration := 1 * time.Minute
+	maxBackoffDuration := time.Duration(maximumCount) * backoffDuration
+	opts := IndexOptions{RepoID: 1, Name: "foo"}
+
+	backoff := backoff{
+		backoffDuration: backoffDuration,
+		maxBackoff:      maxBackoffDuration,
+	}
+
+	now := time.Now()
+	expectedFailuresCount := 0
+
+	// consecutive failures count increments per failure
+	for i := 0; i < maximumCount; i++ {
+		backoff.Fail(now.Add(time.Duration(i)*backoffDuration), logtest.Scoped(t), opts)
+		expectedFailuresCount++
+		AssertFailuresCount(t, expectedFailuresCount, backoff)
+	}
+
+	// consecutive failures count does not change
+	for i := maximumCount - 1; i < failedCount; i++ {
+		backoff.Fail(now.Add(time.Duration(i)*backoffDuration), logtest.Scoped(t), opts)
+		AssertFailuresCount(t, expectedFailuresCount, backoff)
+	}
+}
+
+func TestBackoff_ResetConsecutiveFailures(t *testing.T) {
+	failedCount := 3
+	backoffDuration := 10 * time.Minute
+	maxBackoffDuration := time.Duration(failedCount) * backoffDuration
+	opts := IndexOptions{RepoID: 1, Name: "foo"}
+
+	backoff := backoff{
+		backoffDuration: backoffDuration,
+		maxBackoff:      maxBackoffDuration,
+	}
+
+	for i := 0; i < failedCount; i++ {
+		now := time.Now()
+
+		// fail j consecutive times
+		for j := i; j <= i; j++ {
+			backoff.Fail(now.Add(time.Duration(j)*backoffDuration), logtest.Scoped(t), opts)
+		}
+
+		// reset behavior is independent of current consecutiveFailures count
+		backoff.Reset()
+		AssertFailuresCount(t, 0, backoff)
+	}
+}
+
+func AssertAllow(t *testing.T, now time.Time, b backoff) {
+	if indexingAllowed := b.Allow(now); !indexingAllowed {
+		t.Errorf("Indexing is not allowed to proceed by default at %s due to backing off until %s",
+			now, b.backoffUntil)
+	}
+}
+
+func AssertDisallow(t *testing.T, now time.Time, b backoff) {
+	if indexingAllowed := b.Allow(now); indexingAllowed {
+		t.Errorf("Indexing is allowed to proceed at %s after failure despite being set to backoff until %s",
+			now, b.backoffUntil)
+	}
+}
+
+func AssertFailuresCount(t *testing.T, expected int, b backoff) {
+	if failuresCount := b.consecutiveFailures; failuresCount != expected {
+		t.Errorf("Item currently tracks %d consecutive failures when expected consecutive failures count is %d",
+			failuresCount, expected)
+	}
+}
+
+func EmptyQueue(q *Queue) {
```

Comment:
I don't think we need to export this, do we? Same for the other helpers above.

Reply:
You are correct, I've made this change now